plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/internal/core/server/debug/OpenShiftDebugMode.java

```diff
@@ -155,25 +161,114 @@ public OpenShiftDebugMode disableDevmode() {
 	 * @throws CoreException
 	 */
 	public OpenShiftDebugMode execute(IProgressMonitor monitor) throws CoreException {
-		IDeploymentConfig dc = getDeploymentConfig(context, monitor);
+		Connection connection = OpenShiftServerUtils.getConnection(context.getServer());
+		IResource resource = OpenShiftServerUtils.getResource(context.getServer(), monitor);
+		IDeploymentConfig dc = getDeploymentConfig(resource, connection, monitor);
 		if (dc == null) {
-			throw new CoreException(StatusFactory.errorStatus(
-					OpenShiftCoreActivator.PLUGIN_ID, "No deployment config present that can be updated."));
+			throw new CoreException(OpenShiftCoreActivator.statusFactory().errorStatus(
+		            NLS.bind("Could not find deployment config for resource {0}. "
+		                    + "Your build might be still running and pods not created yet or "
+		                    + "there might be no labels on your pods pointing to the wanted deployment config.", 
+							resource != null? resource.getName() : "")));
 		}
 
-		IPod pod = getNewPod(monitor, dc, ConnectionsRegistryUtil.getConnectionFor(dc));
+		boolean dcUpdated = updateDc(dc, connection, monitor);
+		IPod pod = getPod(dcUpdated, dc, connection, monitor);
 		context.setPod(pod);
 
+		toggleTimeout(resource, connection, context, monitor);
 		toggleDebugger(context, monitor);
 		
 		return this;
 	}
 
-	protected IPod getNewPod(IProgressMonitor monitor, IDeploymentConfig dc, Connection connection) throws CoreException {
-		IPod pod;
-		if (updateDebugmode(dc, context, monitor)
+	private void toggleTimeout(IResource resource, Connection connection, DebugContext context, IProgressMonitor monitor) {
+		if (context.isDebugEnabled()) {
+			setRouteTimeout(resource, connection, monitor);
+		} else {
+			clearRouteTimeout(resource, connection, monitor);
+		}
+	}
+
+	private void setRouteTimeout(IResource resource, Connection connection, IProgressMonitor monitor) {
+		monitor.subTask("Increasing route timeout while debugging...");
+
+		IRoute route = getRoute(resource, connection, monitor);
+		if (route == null) {
+			OpenShiftCoreActivator.pluginLog().logInfo(
+					NLS.bind("Could not increase timeout for debugging: could not find any route for resource {0}", resource.getName()));
+			return;
+		}
+
+		OpenShiftCoreActivator.pluginLog().logInfo(
+				NLS.bind("Setting haproxy timeout for route {0}", route.getName()));
+		
+		if (!hasTimeoutAnnotation(route)) {
```

Comment:
What if the deployment already has a timeout but it is a low one. I think we should back it up in one of our own annotation so that we can recover from several invocations after JBoss Tools restarts